src/__tests__/sharing.js

```diff
@@ -0,0 +1,522 @@
+
+    const expected = {
+        ...alfredUser,
+        analyses: [
+            {
+                analysis_id: "07a3e96e-c21c-11ea-aa22-008cfa5ae621",
+            },
+        ],
+        apps: [
+            {
+                app_id: "e54bfc1a-f811-11e8-8a14-008cfa5ae621",
+                system_id: "de",
+            },
+        ],
+        paths: [
+            {
+                path: "/iplant/home/aramsey/CORE-9077-path.list",
+                permission: "read",
+            },
+            {
+                path: "/iplant/home/aramsey/TestFolder",
+                permission: "read",
+            },
+        ],
+        tools: [
+            {
+                tool_id: "2609f8a8-e928-11e9-a2ea-008cfa5ae621",
+            },
+        ],
+    };
+    expect(result).toStrictEqual(expected);
+});
+
+/**
+ * A user who already has all resources should have no changes when trying to
+ * add missing resources
+ */
+it("tests adding missing resources to a user with all resources", () => {
+    const user = userMap["alfred"];
+    const result = addMissingResourcesToUser(user, resources);
+
+    expect(result).toStrictEqual(user);
+});
+
+it("tests sharing updates with no changes", () => {
+    const user = {
+        alfred: {
+            ...alfredUser,
+            displayPermission: "own",
+            originalPermission: "own",
+            apps: [
+                {
+                    app_id: "e54bfc1a-f811-11e8-8a14-008cfa5ae621",
+                    permission: "own",
+                    system_id: "de",
+                },
+            ],
+        },
+    };
+
+    const sharing = getSharingUpdates(user);
+    const unsharing = getSharingUpdates(userMap, userMap);
+
+    const expectedSharing = {};
+    const expectedUnsharing = {};
+    expect(sharing).toStrictEqual(expectedSharing);
+    expect(unsharing).toStrictEqual(expectedUnsharing);
+});
+
+it("tests sharing updates with data change", () => {
+    const user = {
+        alfred: {
+            ...alfredUser,
+            displayPermission: "own",
+            originalPermission: "read",
+            paths: [
+                {
+                    path: "/iplant/home/aramsey/CORE-9077-path.list",
+                    permission: "read",
+                },
+            ],
+        },
+    };
+
+    const sharingResult = getSharingUpdates(user);
+    const unsharingResult = getUnsharingUpdates(user, {});
+
+    const expectedSharing = {
+        paths: [
+            {
+                paths: [
+                    {
+                        path: "/iplant/home/aramsey/CORE-9077-path.list",
+                        permission: "own",
+                    },
+                ],
+                user: "alfred",
+            },
+        ],
+    };
+
+    const expectedUnsharing = {
+        paths: [
+            {
+                paths: ["/iplant/home/aramsey/CORE-9077-path.list"],
+                user: "alfred",
+            },
+        ],
+    };
```

Comment:
Ah, this might not be super readable/recognizable, but on line 335 we're making a call to the `getUnsharingUpdates` function and we're sending an original `userMap` that contained alfred's permissions (I guess the confusing part is that it says we're updating the permission from `read` to `own`) and then a new `userMap` which is just an empty object `{}`. 

```const unsharingResult = getUnsharingUpdates(user, {});```

The idea is that originally alfred had some permissions, we don't really care which, and now, according to the second `userMap` which is empty, that user has been removed.  It was a quick and dirty way I could think of to test unsharing with the same user object I'm using for sharing.

Reply:
Ah, OK. I guess I didn't read this test carefully enough the first time.

I see now that it's really testing 2 things, with `getUnsharingUpdates` being tested 2nd by `expect(unsharingResult).toStrictEqual(expectedUnsharing)`.